watcher_fsevents_cgo.go

```diff
@@ -160,39 +129,35 @@ func newStream(path string, fn streamFunc) *stream {
 	}
 }
 
-// Start creates a FSEventStream for the given path and schedules it with
-// global runloop. It's a nop if the stream was already started.
+// Start creates a FSEventStream for the given path and schedules on the global dispatch queue.
+// It's a nop if the stream was already started.
 func (s *stream) Start() error {
 	if s.ref != nilstream {
 		return nil
 	}
-	wg.Wait()
 	p := C.CFStringCreateWithCStringNoCopy(C.kCFAllocatorDefault, C.CString(s.path), C.kCFStringEncodingUTF8, C.kCFAllocatorDefault)
 	path := C.CFArrayCreate(C.kCFAllocatorDefault, (*unsafe.Pointer)(unsafe.Pointer(&p)), 1, nil)
 	ctx := C.FSEventStreamContext{}
 	ref := C.EventStreamCreate(&ctx, C.uintptr_t(s.info), path, C.FSEventStreamEventId(atomic.LoadUint64(&since)), latency, flags)
 	if ref == nilstream {
 		return errCreate
 	}
-	C.FSEventStreamScheduleWithRunLoop(ref, runloop, C.kCFRunLoopDefaultMode)
+	C.FSEventStreamSetDispatchQueue(ref, q)
 	if C.FSEventStreamStart(ref) == C.Boolean(0) {
 		C.FSEventStreamInvalidate(ref)
 		return errStart
 	}
-	C.CFRunLoopWakeUp(runloop)
 	s.ref = ref
 	return nil
 }
 
-// Stop stops underlying FSEventStream and unregisters it from global runloop.
+// Stop stops underlying FSEventStream and unregisters it from the global dispatch queue.
 func (s *stream) Stop() {
 	if s.ref == nilstream {
 		return
 	}
-	wg.Wait()
 	C.FSEventStreamStop(s.ref)
```

Comment:
we start the stream with `FSEventStreamStart` at L146, so we stop it here with `FSEventStreamStop` at L159

since we call `FSEventStreamSetDispatchQueue(ref, q)` at L145, maybe it would make sense to call `FSEventStreamSetDispatchQueue(ref, nil)` here?

I tested the change and it seems to work

Reply:
Good question. I looked again at [the docs](https://developer.apple.com/documentation/coreservices/1444164-fseventstreamsetdispatchqueue?language=objc), and they say:

> To remove the stream from the queue on which it was scheduled, call `FSEventStreamSetDispatchQueue` with a NULL queue parameter or call `FSEventStreamInvalidate` which does the same thing.

We call `FSEventStreamInvalidate` on the next line, so I think we're good here.